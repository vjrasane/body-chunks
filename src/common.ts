import { isUndefined } from "lodash";

type Listen = (progress: Progress) => void;

type Progress = {
  bytes: {
    received: number;
    total: number;
  };
  chunks: number[];
};

type OutputValue = string | Buffer | Uint8Array;

type Output = OutputValue[];

export { Listen, Progress, OutputValue, Output };
