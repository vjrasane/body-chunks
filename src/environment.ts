import { isBrowser as browser, isNode as node } from "browser-or-node";

const isBrowser = (): boolean => browser;
const isNode = (): boolean => node;

export { isBrowser, isNode };
