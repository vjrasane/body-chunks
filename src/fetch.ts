import {
  RequestInit as NodeRequestInit,
  default as nodeFetch,
} from "node-fetch";
import { isBrowser } from "./common";

const __fetch = (url: string, options: RequestInit | NodeRequestInit = {}) =>
  isBrowser
    ? fetch(url, <RequestInit>options)
    : nodeFetch(url, <NodeRequestInit>options);

export default __fetch;
