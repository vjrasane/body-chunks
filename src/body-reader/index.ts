import BrowserBodyReader from "./browser";
import NodeBodyReader from "./node";
import { BodyReader } from "./reader";
import {
  RequestInit as NodeRequestInit,
  default as nodeFetch,
} from "node-fetch";
import { isNode, isBrowser } from "../environment";

const getBodyReader = async (
  url: string,
  options: RequestInit | NodeRequestInit = {}
): Promise<BodyReader> => {
  if (isBrowser())
    return new BrowserBodyReader(await fetch(url, <RequestInit>options));
  if (isNode())
    return new NodeBodyReader(await nodeFetch(url, <NodeRequestInit>options));
  throw new Error(
    "Unsupported JavaScript environment, expected one of: node, browser"
  );
};

export { getBodyReader, BodyReader };
