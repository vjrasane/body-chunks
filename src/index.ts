import { getBodyReader, BodyReader } from "./body-reader";

export { BodyReader };
export default getBodyReader;
