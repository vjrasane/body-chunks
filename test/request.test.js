import get from "../src";
import createError from "http-errors";
import { isNode, isBrowser } from "../src/environment";
import BrowserBodyReader from "../src/body-reader/browser";

jest.mock("../src/environment", () => ({
  isNode: jest.fn().mockImplementation(() => false),
  isBrowser: jest.fn().mockImplementation(() => true),
}));

describe("request", () => {
  beforeAll(() => {
    global.fetch = jest.fn();
  })

  beforeEach(() => {
    global.fetch.mockClear();
  });

  it("rejects if request is not ok", async () => {
    global.fetch.mockImplementationOnce(() => ({
      ok: false,
      status: 420,
      statusText: "status",
    }));

    const promise = get("/");

    await expect(promise).rejects.toThrowErrorMatchingSnapshot();
  });

  it("passes request init to fetch", async () => {
    const response = {
      ok: true,
      body: {
        getReader: jest.fn(),
      },
      headers: new Headers({
        "Content-Length": 100,
      }),
    };

    global.fetch.mockImplementationOnce(() => response);

    const requestInit = {};
    const promise = get("/url", requestInit);

    const {
      calls: [[url, opts]],
    } = global.fetch.mock;
    expect(url).toEqual("/url");
    expect(opts).toBe(requestInit);

    await expect(promise).resolves.toBeInstanceOf(BrowserBodyReader);
  });

  it("throws error in unknown environment", async () => {
    isBrowser.mockImplementationOnce(() => false);

    const promise = get("/url");

    await expect(promise).rejects.toThrowErrorMatchingSnapshot();
  });

  afterAll(() => {
    global.fetch = undefined;
  })
});
