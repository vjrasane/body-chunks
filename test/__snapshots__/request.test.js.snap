// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request passes request init to fetch 1`] = `
BrowserBodyReader {
  "__read": [Function],
  "broadcast": [Function],
  "chunks": Array [],
  "closed": false,
  "onProgress": [Function],
  "progress": [Function],
  "progressBus": Bus {
    "_isEventStream": true,
    "_isObservable": true,
    "desc": Desc {
      "_isDesc": true,
      "args": Array [],
      "context": "Bacon",
      "method": "Bus",
    },
    "dispatcher": Dispatcher {
      "_handleEvent": undefined,
      "_subscribe": [Function],
      "ended": false,
      "handleEvent": [Function],
      "observable": [Circular],
      "prevError": undefined,
      "pushing": false,
      "queue": Array [],
      "subscribe": [Function],
      "subscriptions": Array [],
      "unsubSrc": undefined,
    },
    "ended": false,
    "id": 2,
    "initialDesc": Desc {
      "_isDesc": true,
      "args": Array [],
      "context": "Bacon",
      "method": "Bus",
    },
    "pushQueue": undefined,
    "pushing": false,
    "subscriptions": Array [],
    "unsubAll": [Function],
  },
  "read": [Function],
  "readChunk": [Function],
  "reader": undefined,
  "received": 0,
  "total": 100,
}
`;

exports[`request rejects if request is not ok 1`] = `"status"`;

exports[`request throws error in unknown environment 1`] = `"Unsupported JavaScript environment, expected one of: node, browser"`;
